flink-runtime/src/main/java/org/apache/flink/runtime/source/coordinator/SourceCoordinator.java

```diff
@@ -631,8 +631,9 @@ private void ensureStarted() {
         }
     }
 
-    private static class WatermarkAggregator<T> {
+    static class WatermarkAggregator<T> {
         private final Map<T, Watermark> watermarks = new HashMap<>();
+        /** The aggregatedWatermark is the smallest watermark of all keys. */
```

Comment:
nit: I think the style is to use single line comments `//`. The multiline is reserved for signatures, classes, etc.

Reply:
I was thinking that there's also further optimization on the data structure. The WatermarkAggregator only needs to return the min watermark and the set of tracked subtasks. So you could also use a TreeMap.

But inserting an element is now more costly and it's probably safe to assume watermark updates are more frequent than searching for the min watermark.

Reply:
> nit: I think the style is to use single line comments //. The multiline is reserved for signatures, classes, etc.

I see the `/** */` is used at field level in the too many flink code, such as : [PipelinedSubpartition](https://github.com/apache/flink/blob/e35f33f539b09b2c18fb826a144c08541e148dd0/flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/PipelinedSubpartition.java#L87)

Reply:
> I was thinking that there's also further optimization on the data structure. The WatermarkAggregator only needs to return the min watermark and the set of tracked subtasks. So you could also use a TreeMap.

TreeMap orders all elements based on the key instead of value. WatermarkAggregator wants to find the min watermark, it's the value. So I don't know why `TreeMap` is suitable here.

Reply:
Hi @pnowojski  @StefanRRichter , thanks for your review and the great suggestion.

Your suggestion is very helpful, and it's a more general and understandable solution. I have tried it with Flink's `HeapPriorityQueue` and Java's `PriorityQueue`, and I found the performance of `HeapPriorityQueue` is 20~30 times better than Java's `PriorityQueue`.

The following is the benchmark duration:

- The duration based on master branch is 1 min 44s
- The duration based on master branch with old PR is 469ms
- The duration based on master branch with current PR(Flink's `HeapPriorityQueue`) is 284ms
- The duration based on master branch with current PR(Java's `PriorityQueue`) is about 6 s

So the `HeapPriorityQueue` is very strong and I will learn it later, thanks to @StefanRRichter for the great job.

I choose the Flink's `HeapPriorityQueue` based on the benchmark test, and updated this PR.

Reply:
@1996fanrui happy to hear that the suggestion works well. `HeapPriorityQueue` is very similar to Java's implementation. The one big improvement it has is that `remove(...)` is no longer an O(n) operation like for "normal" PQs but O(log n) because we remember the index in the element.

Reply:
Nice! I forgot TreeMap wasn't O(1) lookup--this approach makes sense.